src/psyclone/psyad/transformations/adjoint_trans.py

```diff
@@ -0,0 +1,90 @@
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Authors: R. W. Ford and A. R. Porter, STFC Daresbury Lab
+
+'''This module contains an abstract parent class for adjoint
+transformations.
+
+'''
+from psyclone.psyGen import Transformation
+from psyclone.psyir.backend.fortran import FortranWriter
+from psyclone.psyir.backend.visitor import PSyIRVisitor
+from psyclone.psyir.symbols import DataSymbol
+from psyclone.psyir.transformations import TransformationError
+
+# pylint: disable=abstract-method
+# pylint: disable=super-with-arguments
+
+
+class AdjointTransformation(Transformation):
+    '''An abstract class for Adjoint transformations. Requires a list of
+    active variables to be passed when creating an instance of the
+    class. Also supports an optional writer argument.
+
+    :param active_variables: a list of names of the active variables.
+    :type active_variables: list of str
+    :param writer: the writer to use when outputting error \
+        information. Defaults to FortranWriter.
+    :type writer: subclass of \
+        :py:class:`psyclone.psyir.backend.visitor.PSyIRVisitor`
+
+    :raises TransformationError: if the active_variables or writer \
+        arguments are of the wrong type.
+
+    '''
+    def __init__(self, active_variables, writer=FortranWriter()):
+        super(AdjointTransformation, self).__init__()
+
+        if not isinstance(active_variables, list):
+            raise TransformationError(
+                "The active variables argument should be a list, but found "
+                "'{0}'.".format(type(active_variables).__name__))
+
+        for active_variable in active_variables:
+            if not isinstance(active_variable, DataSymbol):
+                raise TransformationError(
+                    "Active variables should be of type DataSymbol, but "
+                    "found '{0}'.".format(type(active_variable).__name__))
+
+        if not isinstance(writer, PSyIRVisitor):
+            raise TransformationError(
+                "The writer argument should be a PSyIRVisitor but found "
+                "'{0}'.".format(type(writer).__name__))
+
+        # A list of active variables.
+        self._active_variables = active_variables
+        # The writer to use when outputting error information.
+        self._writer = writer
+
+
+__all__ = ["AdjointTransformation"]
```

Comment:
I've followed the usual approach, but I admit the one used for ``__init__`` seems wrong. We are using ``__all__`` because of autoapi. The fact that they are made available here is why we add them to ``__all__``, it is not as a result of added them to ``__all__``. At least that is my understanding. For the regular (none ``__init__`` files) I've used the same words as for PSyclone as I presume the psyclone-ref will pick psyad calls up too.